vhdbuilder/prefetch/pkg/cni/cni.go

```diff
@@ -0,0 +1,72 @@
+package cni
+
+import (
+	"bytes"
+	_ "embed"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/Azure/agentbaker/vhdbuilder/prefetch/pkg/component"
+)
+
+var (
+	//go:embed cni-prefetch.sh.gtpl
+	prefetchScript         string
+	prefetchScriptTemplate = template.Must(template.New("cniprefetch").Parse(prefetchScript))
+
+	// this represents the CNI-related container images and their respective binaries
+	// which are in-scope for prefetch optimization
+	prefetchList = CNIPrefetchList{
+		"mcr.microsoft.com/containernetworking/cni-dropgz:*": []string{"dropgz"},
+		"mcr.microsoft.com/oss/calico/pod2daemon-flexvol:*":  []string{"usr/local/bin/flexvol"},
+		"mcr.microsoft.com/oss/calico/cni:*": []string{
+			"opt/cni/bin/bandwidth",
+			"opt/cni/bin/calico",
+			"opt/cni/bin/calico-ipam",
+			"opt/cni/bin/flannel",
+			"opt/cni/bin/host-local",
+			"opt/cni/bin/install",
+			"opt/cni/bin/loopback",
+			"opt/cni/bin/portmap",
+			"opt/cni/bin/tuning",
+		},
+	}
+)
+
+// Generate generates and saves the CNI prefetch script to disk based on the specified component list
+// and destination path. This can later be generalized to an interface if we need to prefetch
+// other non-CNI related images/binaries.
```

Comment:
apart from the binaries listed in `prefetchList`, is there anything CNI specific here? I wonder if it's worth removing the "cni" names in pkgs/structs/functions to make it easier if we ever want to use this for other components.

Reply:
yeah that's fair, my idea was to put different high-level things we want to prefetch, such as CNI in this case, into separate pkgs that are all self-contained and provide respective `Generate` functions which get called by the main pkg - though you're probably right in that this pkg could be further generalized to "prefetching binaries from a container image"